kdf.js

```diff
@@ -0,0 +1,181 @@
+  }
+
+  crypto_kdf_keygen(key)
+  return key
+}
+
+/**
+ * Initializes key derivation.
+ *
+ * @public
+ * @param {Buffer} buffer
+ * @param {Buffer} key
+ * @return {Object}
+ * @throws TypeError
+ */
+function init(buffer, key) {
+  if (buffer) {
+    if (false === isBuffer(buffer)) {
+      throw new TypeError('kdf.init: Expecting context to be a buffer.')
+    }
+
+    if (crypto_kdf_CONTEXTBYTES !== buffer.length) {
+      throw new TypeError(`kdf.init: Invalid context length: ${buffer.length}`)
+    }
+  }
+
+  if (key && false === isBuffer(key)) {
+    throw new TypeError('kdf.init: Expecting key to be a buffer.')
+  }
+
+  if (undefined === key) {
+    throw new TypeError('kdf.init: Expecting key to be defined.')
+  }
+
+  if (crypto_kdf_KEYBYTES !== key.length) {
+    throw new TypeError(`kdf.init: Invalid key length: ${key.length}`)
+  }
+
+  return {
+    buffer: buffer || Buffer.allocUnsafe(crypto_kdf_CONTEXTBYTES),
+    subkey: null,
+    key
+  }
+}
+
+/**
+ * Updates the subkey in the context object.
+ *
+ * @public
+ * @param {Object} ctx
+ * @param {Number} id
+ * @return {Buffer}
+ * @throws TypeError
+ */
+function update(ctx, id) {
+  if ('object' !== typeof ctx) {
+    throw new TypeError('kdf.update: Expecting ctx to be an object.')
+  }
+
+  if (ctx.subkey && !isBuffer(ctx.subkey)) {
+    throw new TypeError('kdf.update: Expecting ctx.subkey to be a buffer.')
+  }
+
+  if (!isBuffer(ctx.buffer)) {
+    throw new TypeError('kdf.update: Expecting ctx.buffer to be a buffer.')
+  }
+
+  if (crypto_kdf_CONTEXTBYTES !== ctx.buffer.length) {
+    throw new TypeError(`kdf.update: Invalid buffer length: ${ctx.buffer.length}.`)
+  }
+
+  if (!isBuffer(ctx.key)) {
+    throw new TypeError('kdf.update: Expecting ctx.key to be a buffer.')
+  }
+
+  if (crypto_kdf_KEYBYTES !== ctx.key.length) {
+    throw new TypeError(`kdf.update: Invalid buffer length: ${ctx.key.length}.`)
+  }
+
+  if ('number' !== typeof id) {
+    throw new TypeError('kdf.update: Expecting id to be a number.')
+  }
+
+  if (id < 0 || id > (2 ** 64) - 1) {
+    throw new TypeError('kdf.update: Expecting id to be between 0 and (2^64)-1.')
+  }
+
+  ctx.subkey = ctx.subkey || Buffer.allocUnsafe(crypto_kdf_KEYBYTES)
+  crypto_kdf_derive_from_key(ctx.subkey, id, ctx.buffer, ctx.key)
+
+  return ctx.subkey
+}
+
+/**
+ * Final step to null the original context subkey.
+ *
+ * @public
+ * @param {Object} ctx
+ * @return {Buffer}
+ * @throws TypeError
+ */
+function final(ctx) {
+  if ('object' !== typeof ctx) {
+    throw new TypeError('kdf.final: Expecting ctx to be an object.')
+  }
+
+  const { subkey } = ctx
+  ctx.subkey = null
+
+  return subkey
+}
+
+/**
+ * Derives a subkey using the master key and context.
+ *
+ * @public
+ * @param {Buffer} buffer
+ * @param {Buffer} key
+ * @param {Number} iterations
+ * @return {Buffer}
+ * @throws TypeError
+ */
+function derive(buffer, key, iterations) {
```

Comment:
This function requires the caller to supply a buffer or falsy argument. The caller should only have to supply the key and iterations and potentially, an optional buffer, which should be in the last argument position.